src/extension.ts

```diff
@@ -4,148 +4,194 @@
+                                console.error(`stderr: ${data}`);
+                            });
+
+                            childProcess.on('close', (code) => {
+                                console.log(`LSP exited with code ${code}`);
+                            });
+                        } else {
+                            console.log(`Could not resolve smithy-language-server implementation`)
+                        }
+                    })
+
+                    // Send raw output to a file
+                    if (!fs.existsSync(context.storagePath))
+                        fs.mkdirSync(context.storagePath);
+                });
+            });
+        };
+
+        return getCoursier(context.extensionPath)
+            .then(binaryPath => startServer(binaryPath))
+    };
+
+    // Options to control the language client
+    let clientOptions: LanguageClientOptions = {
+        // Register the server for plain text documents
+        documentSelector: [{ scheme: 'file', language: 'smithy' }, { scheme: 'smithyjar', language: 'smithy' }],
+        synchronize: {
+            // Notify the server about file changes to 'smithy-build.json' or 'smithy.json' files contained in the workspace
+            fileEvents: workspace.createFileSystemWatcher('**/{smithy-build}.json')
+        }
+    };
+
+    // Create the language client and start the client.
+
+    client = new LanguageClient(
+        'smithyLsp',
+        'Smithy LSP',
+        createServer,
+        clientOptions
+    );
+    const smithyContentProvider = createSmithyContentProvider(client);
+    context.subscriptions.push(workspace.registerTextDocumentContentProvider('smithyjar', smithyContentProvider));
+
+    // Start the client. This will also launch the server
+    client.start();
 }
 
 export function deactivate(): Thenable<void> | undefined {
-	if (!client) {
-		return undefined;
-	}
-	return client.stop();
+    if (!client) {
+        return undefined;
+    }
+    return client.stop();
 }
 
 function createSmithyContentProvider(languageClient: LanguageClient): vscode.TextDocumentContentProvider {
-	return <vscode.TextDocumentContentProvider>{
-		provideTextDocumentContent: async (uri: vscode.Uri, token: CancellationToken): Promise<string> => {
-			return languageClient.sendRequest(ClassFileContentsRequest.type, { uri: uri.toString() }, token).then((v: string): string => {
-				return v || '';
-			});
-		}
-	}
+    return <vscode.TextDocumentContentProvider>{
+        provideTextDocumentContent: async (uri: vscode.Uri, token: CancellationToken): Promise<string> => {
+            return languageClient.sendRequest(ClassFileContentsRequest.type, { uri: uri.toString() }, token).then((v: string): string => {
+                return v || '';
+            });
+        }
+    }
+}
+
+function getCoursier(extensionPath: string): Promise<string> {
+    function binPath(filename: string) { 
+        return path.join(extensionPath, filename);

```

Comment:
**suggestion:** In some situations (pretty rare, because I think it'll only impact users of nix/nixos), `extensionPath` is a read-only directory. Can we check for that and fallback to some other location, maybe `$TMPDIR` (at least on unix systems)?